third_party/tlx/dialect/lib/Transforms/ResolvePlaceholderLayouts.cpp

```diff
@@ -0,0 +1,215 @@
+#include "IR/Dialect.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/IR/Types.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+#include "triton/Dialect/TritonNvidiaGPU/IR/Dialect.h"
+#include "llvm/Support/Debug.h"
+#include <numeric>
+
+#define DEBUG_TYPE "tlx-resolve-placeholder-layouts"
+#define DBGS() (llvm::errs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) DBGS() << X << "\n"
+
+using namespace mlir;
+namespace ttg = ::mlir::triton::gpu;
+namespace ttng = ::mlir::triton::nvidia_gpu;
+
+namespace mlir {
+namespace triton {
+namespace tlx {
+
+#define GEN_PASS_DEF_TLXRESOLVEPLACEHOLDERLAYOUTS
+
+#include "tlx/dialect/include/Transforms/Passes.h.inc"
+
+/// Check if an attribute is any of the dummy layout types
+static bool isDummyLayoutAttr(Attribute attr) {
+  return isa<DummyRegisterLayoutAttr>(attr);
+}
+
+/// Extract the dummy layout attribute from a type, if present
+static Attribute getDummyLayoutFromType(Type type) {
+  if (auto tensorType = dyn_cast<RankedTensorType>(type)) {
+    if (auto encoding = tensorType.getEncoding()) {
+      if (isDummyLayoutAttr(encoding))
+        return encoding;
+    }
+  }
+  if (auto memDescType = dyn_cast<ttg::MemDescType>(type)) {
```

Comment:
This could be omitted for now, as we are targeting register layouts only?

Reply:
Yes this is a copy over.